feat(topology): add permissions to topology plugin

PatAKnight opened the pull request:
## Description

This PR adds permissions to the Topology plugin. These permissions, `topology.panel.read` and `topology.view.read`, to restrict access to the Topology plugin. Permission `topology.view.read` will limit access to the Topology view, see below. Permission `topology.panel.read` will limit access to the Topology sidebar.

# Fixes

- Fixes: [RHIDP-1794](https://issues.redhat.com/browse/RHIDP-1794)

## Screenshots

`topology.view.read`
![Screenshot from 2024-05-15 11-16-17](https://github.com/janus-idp/backstage-plugins/assets/82788581/5ee18543-51a2-45da-b0fc-ba29a057cf4d)
allow

![Screenshot from 2024-05-15 11-16-30](https://github.com/janus-idp/backstage-plugins/assets/82788581/f7262883-870f-413c-a6d5-99ecca253847)
deny (used `TopologyEmptyState` because the approach is similar to `catalog.entity.read` showing an empty state when navigating to the catalog when set to deny)

`topology.panel.read`
![Screenshot from 2024-05-15 11-17-33](https://github.com/janus-idp/backstage-plugins/assets/82788581/8df3613c-ba65-4f50-bf10-63d3c31ffe57)
allow (details tab)

![Screenshot from 2024-05-15 11-17-35](https://github.com/janus-idp/backstage-plugins/assets/82788581/45b9be45-752c-4e16-a7f4-f6fa8843bf5e)
allow (resources tab)

![Screenshot from 2024-05-15 11-16-49](https://github.com/janus-idp/backstage-plugins/assets/82788581/b9c5e196-f8d2-4ec9-9e1f-3eb687b7cf3d)
deny (tabs are disabled) (went with the `EmptyState` from patternfly and passed in the text `Unauthorized: Unable to view data`, didn't know what text to include and whether or not we should use the `TopologyEmptyState`)

`catalog.entity.read`
![Screenshot from 2024-05-15 11-17-33](https://github.com/janus-idp/backstage-plugins/assets/82788581/c759084f-78ea-409f-bea6-2259aff603c3)
allow

![Screenshot from 2024-05-15 11-17-49](https://github.com/janus-idp/backstage-plugins/assets/82788581/03772f4c-b30b-463d-b6b6-df9f3a8f9fcc)
deny (before reload of browser)

![Screenshot from 2024-05-15 11-20-53](https://github.com/janus-idp/backstage-plugins/assets/82788581/98542cef-9968-4b2a-9366-9a87c4e03791)
deny (after reload of browser, view is originating from catalog plugin)

## Special notes for the Review

### How to test

1. Configure the Topology plugin, Kubernetes plugin, Kubernetes Backend plugin
2. Create a policy csv file for the permission policies
    ```csv
    g, user:default/<YOUR_USERNAME>, role:default/topology

    p, role:default/topology, topology.panel.read, read, allow
    p, role:default/topology, topology.view.read, read, allow
    p, role:default/topology, catalog-entity, read, allow
    ```
3. Configure the RBAC Backend plugin
    ```yaml
    permission:
      enabled: true
      rbac:
        policies-csv-file: ../<PATH>/<TO>/<CSV>/rbac-policy.csv
        policyFileReload: true
    ```
4. Ensure auth is configured
5. Ensure that there is a cluster to connect to, a deployment and catalog entity that can be referenced, and a user that you can sign into in the catalog
6. View the Topology side for you deployed catalog entity
7. Change above permission policies to denied and reload to restrict access

This is marked as a draft to get review and sign off on UI / UX / and the RBAC Backend team. I took the approach of getting everything implemented prior to getting review on how we want it to look because of the priority. Most of the changes are probably where we want them. The biggest things that need review are the text and whether we want to use an empty state or something more customized.

@ShiranHi @invincibleJai @divyanshiGupta

ShiranHi commented:
@PatAKnight it looks good. I have three questions regarding the content displayed when permission is denied:

> deny (used `TopologyEmptyState` because the approach is similar to `catalog.entity.read` showing an empty state when navigating to the catalog when set to deny)
Can we change the text to be:
Title: Permission required
Text: To view Topology, contact your administrator to give you catalog.entity.read permission.

> deny (before reload of browser)
Can we change the text in the tab to be "Permission required to view data."
I'm also curious about the image there. Where did it come from? I don't find it helpful 🤔

> deny (after reload of browser, view is originating from catalog plugin)
Is this a default empty state page? Can we improve it to be a full empty state like [this one](https://www.figma.com/design/GeVb7rc390TkxeQNnJIcBD/RHDH-Theme-(pages)?node-id=110%3A933&t=bzYC8X4owP8slqZq-1)?

PatAKnight commented:
> Can we change the text to be: Title: Permission required Text: To view Topology, contact your administrator to give you catalog.entity.read permission.

Yeah, I can either expand upon the `TopologyEmptyState` to include [additional options](https://github.com/PatAKnight/backstage-plugins/blob/2860272909877e31322c5f2276721972eefa28c0/plugins/topology/src/components/Topology/TopologyEmptyState.tsx#L11) for denial text or create a separate `TopologyDeniedState` that covers the change in text. @divyanshiGupta What do you think?

> 
> Can we change the text in the tab to be "Permission required to view data." I'm also curious about the image there. Where did it come from? I don't find it helpful 🤔
>

I had a conversation with @divyanshiGupta the other day in which we agreed to remove the additional permission `topology.panel.read`. That permission was originally to block access to the sidebar that you are seeing there. It is no longer relevant. I can update the PR to remove those images, if that would help.

> Is this a default empty state page? Can we improve it to be a full empty state like [this one](https://www.figma.com/design/GeVb7rc390TkxeQNnJIcBD/RHDH-Theme-(pages)?node-id=110%3A933&t=bzYC8X4owP8slqZq-1)?

Yes, that is the default empty state page after the `catalog-entity` `read` permission has been set to deny. It blocks all access to the resource. And is coming from the catalog plugin

divyanshiGupta commented:
> > Can we change the text to be: Title: Permission required Text: To view Topology, contact your administrator to give you catalog.entity.read permission.
> 
> Yeah, I can either expand upon the `TopologyEmptyState` to include [additional options](https://github.com/PatAKnight/backstage-plugins/blob/2860272909877e31322c5f2276721972eefa28c0/plugins/topology/src/components/Topology/TopologyEmptyState.tsx#L11) for denial text or create a separate `TopologyDeniedState` that covers the change in text. @divyanshiGupta What do you think?
> 

+1 for expanding the `TopologyEmptyState` and making it more generic so that it can be used for both usecases.

> > Can we change the text in the tab to be "Permission required to view data." I'm also curious about the image there. Where did it come from? I don't find it helpful 🤔
> 
> I had a conversation with @divyanshiGupta the other day in which we agreed to remove the additional permission `topology.panel.read`. That permission was originally to block access to the sidebar that you are seeing there. Since I removed it, the image is now out of date. I can update the PR to remove those images, if that would help.
> 

+1

divyanshiGupta commented:
@PatAKnight the kubernetes plugin has a permission that can be setup for limiting access for viewing pod logs. Users can also see pod logs from topology node sidebar. So should we have a similar permission here as well?

PatAKnight commented:
@divyanshiGupta We are actually already utilizing the permission from the Kubernetes plugin. The `kubernetesProxyApi` is already protected by the Kubernetes proxy permission [here](https://github.com/janus-idp/backstage-plugins/blob/main/plugins/topology/src/hooks/usePodLogs.ts#L25). But I will go in and add it to the docs in the Topology plugin and the RBAC plugin to make sure users are aware that the Topology plugin uses that permission as well.

PatAKnight commented:
Updated screenshot
![Screenshot from 2024-05-21 08-49-21](https://github.com/janus-idp/backstage-plugins/assets/82788581/1af97686-8ce4-4aed-a526-d775521f8800)

And updated the docs with the inclusion of `kubernetes.proxy, use` permission

divyanshiGupta commented on plugins/topology/src/components/Topology/TopologyViewWorkloadComponent.tsx:
```diff
@@ -126,8 +140,15 @@ const TopologyViewWorkloadComponent = ({
+            {isDataModelEmpty ||
+            !(
+              topologyViewPermissionResult.allowed &&
+              catalogEntityPermissionResult.allowed
```
@PatAKnight should we update documentation to inlcude that the user should also have `catalogEntityReadPermission` to view the topology view?

PatAKnight replied:
Yep, that makes sense to me.

PatAKnight replied:
Updated with the mention of the `catalog-entity, read` permission.

divyanshiGupta reviewed (approved):
LGTM

divyanshiGupta commented on plugins/topology/src/components/Topology/TopologyViewWorkloadComponent.tsx:
```diff
@@ -49,6 +54,15 @@ const TopologyViewWorkloadComponent = ({
+
+  const catalogEntityPermissionResult = usePermission({
+    permission: catalogEntityReadPermission,
+    resourceRef: catalogEntityReadPermission.resourceType,
```
@PatAKnight I removed lgtm so that @AndrienkoAleksandr can also reveiw the PR but it still got merged. :P
Last time I checked with @AndrienkoAleksandr this is not the way catalogEntityReadPermission authorization works as it expects the complete resourceRef like `user:default/div`. I have opened one UI issue regarding this as when conditions are added to a role of admin this /api/permission/authorize returns this error for catalogEntityReadPermission : `
"Entity reference \"catalog-entity\" had missing or empty kind (e.g. did not start with \"component:\" or similar)"`
